src/lib/png/png.c

```diff
@@ -0,0 +1,976 @@
+			}
+			break;
+
+		case 4:
+			if (precon) {
+				for (i = 0; i != bytewidth; i++) {
+					recon[i] = (scanline[i] + precon[i]);
+				}
+				for (i = bytewidth; i < length; i++) {
+					recon[i] = (scanline[i] + predictor(recon[i - bytewidth], precon[i], precon[i - bytewidth]));
+				}
+			} else {
+				for (i = 0; i != bytewidth; i++) {
+					recon[i] = scanline[i];
+				}
+				for (i = bytewidth; i < length; i++) {
+					recon[i] = (scanline[i] + recon[i - bytewidth]);
+				}
+			}
+			break;
+
+		default:
+			return -1; /* error in filter */
+		}
+		return 0;
+}
+
+static int unfilter_first_step(uint8_t *out, const uint8_t *in, int w, int h, int bpp) {
+	int i;
+	uint8_t* prevline = 0;
+
+	int bytewidth = (bpp + 7) / 8;
+	int linebytes = (w * bpp + 7) / 8;
+
+	for (i = 0; i < h; i++) {
+		int outindex = linebytes * i;
+		int inindex = (1 + linebytes) * i;
+		uint8_t filterType = in[inindex];
+
+		if (unfilter_second_step(&out[outindex], &in[inindex + 1],
+			prevline, bytewidth, filterType, linebytes) == -1) {
+				printf("Error in filter\n");
+				return -1;
+			}
+
+		prevline = &out[outindex];
+	}
+	return 0;
+}
+
+/**
+ * 1)This function converts the filtered-padded-interlaced data in img
+ * 2)out: big enough for contain full image,
+ * 3)in: contain the full decompressed data from
+ * the IDAT chunks (with filter index bytes and possible padding bits)
+ * 4)Steps:
+ * a) if no Adam7: 1) unfilter 2) TODO: remove padding bits (if bpp < 8)
+ * b) TODO: if adam7: 1) 7x unfilter 2) 7x remove padding bits 3) Adam7_deinterlace
+ * return 0 - success or error
+*/
+static int after_decode_unfilter(uint8_t *out, uint8_t *in, int w, int h, int interflace, int bpp) {
+	if (interflace == 0) {
+		/* TODO removePaddingBits */
+		unfilter_first_step(out, in, w, h, bpp);
+	} else {
+		printf("TODO : Adam7 not supported yet\n");
+		return -1;
+	}
+	return 0;
+}
+
+/**
+ * this function start of decoding IDAT
+ * out : buffer for decoded image, big enough for img
+ * in : contain IDAT bytes
+ * inlen : length of in(without zlib_head)
+*/
+static int inflate(uint8_t *out, uint8_t *in, int inlen, struct png_header *png_head) {
+	int bp = 3;/* bit point in in */
+	int pos = 0;/*pos in out */
+	int btype = 2;/* huffman type (1 - fixed/2 - dynamic) */
+	int bpp;
+	uint8_t *Huffman_out = (uint8_t*)malloc(png_head->width * png_head->height * 5);
+
+	if (inflate_huffman(Huffman_out, in, &bp, &pos, inlen, btype) != 0) {
+		printf("Error : can't decode IDAT chunk.\n");
+		return -1;
+	}
+
+	bpp = png_head->color_type * png_head->bit_depth;
+
+	if (after_decode_unfilter(out, Huffman_out, png_head->width, png_head->height,
+												png_head->interlace_method, bpp) != 0) {
+		printf("Error while unfilter\n");
+		return -1;
+	}
+	return 0;
+}
+
+static int read_IDAT(int fd, uint8_t *image, int len_IDAT, struct png_header *png_head) {
+	uint8_t zlib[2];
+
+	len_IDAT -= 2; /* IDAT len minus len of zlib header */
+
+	uint8_t *data_IDAT = (uint8_t*)malloc(len_IDAT);
```

Comment:
Please, add check for successful memory allocation.

p.s. Where this memory is being freed?

Reply:
done